telegram/telegram.py

```diff
@@ -0,0 +1,199 @@
+        static=True,
+        required=True,
+    )
+    bot_token = ConfigText(
+        "Our bot's unique token to access the Telegram API",
+        static=True,
+        required=True,
+    )
+    outbound_url = ConfigText(
+        'The URL our bot should make requests to',
+        default='https://api.telegram.org/bot', static=True,
+    )
+    inbound_url = ConfigText(
+        'The URL our transport will listen on for Telegram updates',
+        static=True, required=True,
+    )
+
+
+class TelegramTransport(HttpRpcTransport):
+    """
+    Telegram transport for Vumi
+    """
+    transport_type = 'telegram'
+    transport_name = 'telegram_transport'
+    _requests = {}
+
+    CONFIG_CLASS = TelegramTransportConfig
+
+    @classmethod
+    def agent_factory(cls):
+        """For swapping out the Agent for use in tests"""
+        return Agent(reactor)
+
+    @inlineCallbacks
+    def setup_webhook(self):
+        # NOTE: Telegram currently only supports ports 80, 88, 443 and 8443 for
+        #       webhook setup
+        url = '%s/setWebhook' % self.outbound_url.rstrip('/')
+        http_client = HTTPClient(self.agent_factory())
+
+        try:
+            r = yield http_client.post(
+                url=url,
+                data=json.dumps({'url': self.inbound_url}),
+                headers={'Content-Type': ['application/json']}
+                )
+            content = yield r.content()
+            res = json.loads(content)
+            if r.code == http.OK and res['ok']:
+                log.info('Webhook set up on %s' % self.inbound_url)
+            else:
+                log.info('Webhook setup failed: %s' % res['description'])
+
+        # Treat page redirects as errors, since Telegram seems to redirect us
+        # when our bot token is invalid
+        except ResponseFailed:
+            log.info('Webhook setup failed: Invalid token')
+        # In case we get a response from Telegram that isn't JSON
+        except ValueError:
+            log.info('Webhook setup failed: Unexpected response')
+
+    @inlineCallbacks
+    def setup_transport(self):
+        config = self.get_static_config()
+        self.outbound_url = '%s%s' % (config.outbound_url.rstrip('/'),
+                                      config.bot_token)
+        self.inbound_url = config.inbound_url
+        self.bot_username = config.bot_username
+        self.web_path = config.web_path
+        self.web_port = config.web_port
+        self.clock = self.get_clock()
+        self.rpc_resource = HttpRpcResource(self)
+
+        self.web_resource = yield self.start_web_resources(
+            [
+                (self.rpc_resource, self.web_path),
+            ],
+            self.web_port)
+
+        yield self.setup_webhook()
+
+    def get_clock(self):
+        return reactor
+
+    @inlineCallbacks
+    def handle_raw_inbound_message(self, message_id, request):
+        update = json.loads(request.content.read())
```

Comment:
nitpick which you can feel free to ignore: you can skip the `.read()` if you use `json.load()` as that takes a file as an argument and `request.content` can be treated as such. Quack! Quack!